### PR TITLE
Batch norm

### LINKED ISSUE #22050: batch_norm

### PR DESCRIPTION
**Pull request by Ambro19**
Close #22050

This function applies Batch Normalization as described in the paper Batch Normalization: Accelerating Deep Network Training by Reducing Internal Covariate Shift.

### CODE REVIEW

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Hey @Ambro19 ! Thanks for opening this PR. Would you please uncommit the chanegs you have made inside the `array_api_testing` folder. Morever, the issue that you have linked in the PR is about adding `batch_norm` to the paddle frontend but you have added `instance_norm` as well. Would you please remove all the unrelated changes from this PR? Feel free to request another review when you are done. :)

**Comment by Ambro19** [comment]
Hey @hello-fri-end, I am sorry for the confusion. I will go ahead and working on them. Thanks!

**Comment by Ambro19** [comment]
@hello-fri-end, I think I successfully reverted it. But I am not how to ask for another review :)

**Comment by Ambro19** [comment]
Hi @hello-fri-end, can you please check this out for me: https://github.com/unifyai/ivy/compare/main...Ambro19:ivy:batch_norm?expand=1

**Comment by Ambro19** [comment]
Hey @hello-fri-end, @Felix, @Aarsh, can you please look at my PR when you've chance: https://github.com/unifyai/ivy/pull/22273

File: ivy/functional/frontends/paddle/nn/functional/norm.py [hunk 1]
```diff
@@ -8,3 +8,34 @@
...
+    return ivy.batch_norm(
+        x,
+        running_mean,
+        running_var,
+        weight,
+        bias,
+        training,
+        momentum,
+        epsilon,
+        data_format,
```
Comment (hello-fri-end) [inline ivy/functional/frontends/paddle/nn/functional/norm.py:38]: You need to add some logic to map the `data_format` parameter here with the format that `ivy.batch_norm` expects. Basically, the `data_format` parameter in ivy can be either `NCS` or `NSC` ( channel second or channel last respectively) while in paddle, it can have the following values:
```
“NC”, “NCL”, “NCHW”, “NCDHW”, “NLC”, “NHWC” or “NDHWC”, 
```
You can easily observe that that first 3 map to `NCS` while the last 2 map to `NSC`

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py [hunk 2]
```diff
@@ -43,3 +43,38 @@ def test_paddle_layer_norm(
         bias=offset[0],
         epsilon=eps,
     )
+
+
+# bach_norm
+@handle_frontend_test(
+    fn_tree="paddle.nn.functional.bach_norm",
+    values_tuple=_generate_data_layer_norm(
+        available_dtypes=helpers.get_dtypes("float"),
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py:52]:
```suggestion
        available_dtypes=helpers.get_dtypes("valid"),
```

`valid` automatically returns the datatypes that are supported by the function - if it's only `float` then that's what `valid` will return.

Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py:52]: `layer_norm` does not have the same arguments as `batch_norm`. You'll need to write another composite strategy to generate the data for this function. You should generate a value for each of the arguments in your `st.composite` strategy - you can take a look at the test of `ivy.batch_norm` to get an idea of the same

File: ivy/functional/frontends/paddle/nn/functional/norm.py [hunk 3]
```diff
@@ -8,3 +8,34 @@
 @with_supported_dtypes({"2.5.1 and below": ("float32", "float64")}, "paddle")
 def layer_norm(x, normalized_shape, weight=None, bias=None, epsilon=1e-05, name=None):
     return ivy.layer_norm(x, normalized_shape, weight, bias, epsilon)
+
+
+# batch_norm
+@to_ivy_arrays_and_back
+@with_supported_dtypes({"2.5.1 and below": ("float32", "float64")}, "paddle")
```
Comment (hello-fri-end) [inline ivy/functional/frontends/paddle/nn/functional/norm.py:15]: Out of curiosity, how did you arrive at these `supported_dtypes`, did you create a simple script to test the supported dtypes?

**Review by hello-fri-end — CHANGES_REQUESTED** [review]

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Please resolve the comments left in the previous review first and then request a review. Thanks.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py (hunk 2 above)
Comment (Ambro19) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py]: Ok, let me work on trying to write another composite strategy and I will get back you soon...

**Review by Ambro19 — COMMENTED** [review]
Hi @hello-fri-end, I fixed them as you required with a couple of questions. Please review and let me know. Thanks

File: ivy/functional/frontends/paddle/nn/functional/norm.py (hunk 1 above)
Comment (hello-fri-end) [reply ivy/functional/frontends/paddle/nn/functional/norm.py:38]: Yup, it's the former 
```python
if data_format in [“NC”, “NCL”, “NCHW”, “NCDHW”]: # Map to 4 first values
    return ivy.batch_norm with data_format parameter as "NCS"
elif data_format_paddle in [“NLC”, “NHWC", “NDHWC”]: # Map to last 3 values
return ivy.batch_norm with data_format parameter as "NSC"
else:
raise ValueError('Invalid data_format value')

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py (hunk 2 above)
Comment (hello-fri-end) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py]: Yup.

File: ivy/functional/frontends/paddle/nn/functional/norm.py (hunk 3 above)
Comment (hello-fri-end) [reply ivy/functional/frontends/paddle/nn/functional/norm.py:15]: You can modify the following script to generate the values for the parameters of batch_norm and then call `paddle.bacth_norm` instead of `paddle.numel`:

```python
#, "uint16", "bfloat16"
dtypes_list = ["int8", "int16", "int32", "int64", "uint8", "float16", "float32", "float64", "complex64", "complex128", "bool",]
unsupported = []
supported = []
for dtype in dtypes_list:
    try:
        a = paddle.to_tensor([[[9.]]],dtype=dtype)
        # insert the function you want to test in the following line
        #paddle.normal(a,a)
        paddle.numel(a)
        supported.append(a.dtype.name)
    except:
        # use this to generate native dtype
        unsupported.append(a.dtype.name)
        # use this to generate string dtype
        #unsupported.append(dtype)
print(unsupported)
print(supported)
```

Comment (Ambro19) [reply ivy/functional/frontends/paddle/nn/functional/norm.py:15]: Oh, ok I will modify it. 

I also wrote this function, batch_norm_data to generate data for batch_norm, that I wanted you to check:

# global
from hypothesis import strategies as st

# local
import ivy_tests.test_ivy.helpers as helpers
from ivy_tests.test_ivy.helpers.testing_helpers import handle_frontend_test

from ivy_tests.test_ivy.test_functional.test_nn.test_norms import (
    _generate_data_layer_norm,
)
@st.composite
def batch_norm_data(draw):
    # Define the strategies for each parameter of the batch_norm function
    x_strategy = draw(st.lists(st.floats()))  
    mean_strategy = draw(st.lists(st.floats()))
    var_strategy = draw(st.lists(st.floats())) 
    offset_strategy = draw(st.lists(st.floats())) 
    scale_strategy = draw(st.lists(st.floats()))
    training_strategy = draw(st.floats()) 
    epsilon_strategy = draw(st.floats()) 
    momentum_strategy = draw(st.floats()) 
    data_format_strategy = draw(st.text()) 
    out_strategy = draw(st.lists(st.floats()))

    # Return a tuple of the drawn values
    return x_strategy, mean_strategy, var_strategy, offset_strategy,scale_strategy, training_strategy,epsilon_strategy,momentum_strategy, data_format_strategy, out_strategy 


I don't know if it looks all right. But also I am getting some error running it.

Comment (Ambro19) [reply ivy/functional/frontends/paddle/nn/functional/norm.py:15]: Hi @hello-fri-end, just an update: I worked on it as instructed. But I am running on some errors due to machine (Unable Windows Long Path). I troubleshoot it and get back to you soon.

**Comment by Ambro19** [comment]
Hi @hello-fri-end, this is the dtypes generation function for bath_norm parameters that I wrote (please ignore those unrelated comments ...):

import ivy
import paddle

# List of data types to test
dtypes_list = ["int8", "int16", "int32", "int64", "uint8", "float16", "float32", "float64", "complex64", "complex128", "bool"]

unsupported = []
supported = []

for dtype in dtypes_list:
    try:
        # Create a tensor with the current data type
        #a = ivy.array([[[9.]]], dtype=dtype)
        a = paddle.to_tensor([[[9.]]],dtype=dtype)
        
        # Create mean and variance tensors for batch_norm function
        #mean = ivy.array([[[0.]]], dtype=dtype)
        #variance = ivy.array([[[1.]]], dtype=dtype)
        
        # Test the batch_norm function
        #ivy.batch_norm(a, mean, variance)
        
        paddle.bacth_norm(a)
        # If no error is thrown, add the data type to the supported list
        supported.append(a.dtype.name)
    except:
        # If an error is thrown, add the data type to the unsupported list
        unsupported.append(a.dtype.name)

print("Unsupported data types:", unsupported)
print("Supported data types:", supported)

But after running it, I am facing errors due to the PaddlePaddle installation. 

Do you think you can run it on your side and let me know what you found? Meanwhile, I am troubleshooting my PaddlePaddle installation. Thanks

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Hey @Ambro19, not sure what to review here. No change has been made since the last review and you have not implemented batch_norm yet. Could you please undo the changes you have made to the normalize function and implement batch_norm in the paddle frontend along with the test and then re-request a review?

Incase you're still stuck with the supported dtypes, I checked the documentation and this funciton only supports `float32` and `float64`, so you should add the `with_supported_dtypes` decorator to your function with these data-types. Thanks!

**Comment by Ambro19** [comment]
Hey @hello-fri-end, I'm sorry about that, my fault. 

1. I looked at it, I saw that someone has put it back. That's how it looks like on my end. Please let me know if no one has put it back, I will do it.  
2. I implemented both functions as you requested.
3. But as you'll see, we will need to implement the _generate_data_batch_norm() function which should return a tuple containing the data types, input array, mean, variance, scale, and offset. So, because of that I can't really test the batch_norm function, I think.
4. So if you can open the _generate_data_batch_norm() issue, I can implement it as well. Thank you!

File: ivy/functional/frontends/paddle/nn/functional/norm.py [hunk 4]
```diff
@@ -6,13 +6,23 @@
...
 
 
 @to_ivy_arrays_and_back
 @with_supported_dtypes({"2.5.1 and below": ("float32", "float64")}, "paddle")
-def normalize(x, p=2, axis=1, epsilon=1e-12, name=None):
-    if axis < 0:
-        axis = ivy.get_num_dims(x) + axis
-    return ivy.lp_normalize(x, p=p, axis=axis)
+def layer_norm(x, normalized_shape, weight=None, bias=None, epsilon=1e-05, name=None):
+    return ivy.layer_norm(x, normalized_shape, weight, bias, epsilon)
```
Comment (hello-fri-end) [inline ivy/functional/frontends/paddle/nn/functional/norm.py:28]: Could you undo all the changes you have made to `layer_norm` and `normalize` function and only add `batch_norm` function.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py [hunk 5]
```diff
@@ -10,26 +10,25 @@
 )
 
 
-# layer_norm
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py:13]: Again, let's undo all unrelated changes here and only add a new test for `batch_norm`. Thanks!

**Review by hello-fri-end — CHANGES_REQUESTED** [review]

File: ivy/functional/frontends/paddle/nn/functional/norm.py (hunk 4 above)
Comment (Ambro19) [reply ivy/functional/frontends/paddle/nn/functional/norm.py]: Sorry about that. I see that layer_norm didn't get any change, and still here. I added normalization back. I didnt touch theses functions I justed on batch_norm, but I am sure why they got changed. Please let me know about layer_norm becuase I see it here. Thank you.

File: ivy/functional/frontends/paddle/nn/functional/norm.py [hunk 6]
```diff
@@ -6,13 +6,31 @@
...
+    x,
+    mean,
+    variance,
+    offset=None,
+    scale=None,
+    training=False,
+    epsilon=1e-5,
+    momentum=1e-1,
+    data_format="NSC",
+):
```
Comment (hello-fri-end) [inline ivy/functional/frontends/paddle/nn/functional/norm.py:19]: The frontend function should have the same args as the native paddle batch_norm function 
```
paddle.nn.functional.batch_norm(x, running_mean, running_var, weight, bias, training=False, momentum=0.9, epsilon=1e-05, data_format='NCHW', use_global_stats=None, name=None)
```

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py [hunk 7]
```diff
@@ -10,26 +10,25 @@
...
 )
-def test_paddle_layer_norm(
+def test_paddle_batch_norm(
     *,
     values_tuple,
-    normalized_shape,
     eps,
     test_flags,
     frontend,
     on_device,
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_norm.py:27]: Again let's undo all the changes to unrelated files here first.

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Hey @Ambro19 , as mentioned last time, please undo all unrelated changes from this PR. Feel free to request another review when you are done. Thanks :)
